tools/dds/dds-server/lrs-device-controller.cpp

```diff
@@ -446,7 +467,49 @@ void lrs_device_controller::start_streaming( const json & msg )
     _dds_device_server->start_streaming( realdds_streams_to_start );
     _sensors[sensor_index].open( rs_profiles_to_open );
     _sensors[sensor_index].start( [&]( rs2::frame f ) {
-        _dds_device_server->publish_image( f.get_profile().stream_name(), static_cast< const uint8_t * >( f.get_data() ), f.get_data_size() );
+        const auto & stream_name = f.get_profile().stream_name();
+        auto & streams = _dds_device_server->get_streams();
+
+        auto it = streams.find( stream_name );
+        if( it == streams.end() )
+            throw std::runtime_error( "stream '" +stream_name + "' does not exist" );
+        auto & stream = it->second;
+        if( !stream->is_streaming() )
+            throw std::runtime_error( "stream '" +stream_name + "' is not streaming" );
+
+        stream->publish( static_cast< const uint8_t * >( f.get_data() ), f.get_data_size(), f.get_frame_number() );
+
+        if( _md_enabled )
+        {
+            const auto & metadata_stream_name = f.get_profile().stream_name() + " metadata";
+
+            auto it = streams.find( metadata_stream_name );
+            if( it == streams.end() )
+                throw std::runtime_error( "stream '" + metadata_stream_name + "' does not exist" );
+            auto metadata_stream = std::dynamic_pointer_cast< realdds::dds_metadata_stream_server >( it->second );
+            if( !metadata_stream )
+                throw std::runtime_error( "stream '" + stream_name + "' does not have a matching metadata stream" );
+
+            json metadata = json( {
+                // Special data that is always expected by realdds
+                { "frame_id", std::to_string( f.get_frame_number() ) },
+                { "timestamp", f.get_timestamp() },
+                { "timestamp_domain", f.get_frame_timestamp_domain() }
+                } );
+            if( f.is< rs2::depth_frame >() )
+                metadata["depth_units"] = f.as< rs2::depth_frame >().get_units();
+
+            for( size_t i = 0; i < static_cast< size_t >( RS2_FRAME_METADATA_COUNT ); ++i )
+            {
+                rs2_frame_metadata_value val = static_cast< rs2_frame_metadata_value >( i );
+                if( f.supports_frame_metadata( val ) )
+                {
+                    metadata[rs2_frame_metadata_to_string( val )] = f.get_frame_metadata( val );
```

Comment:
I think you'd want to separate actual metadata from metadata "header" fields -- e.g., the frame-id, timestamp/-domain are all in the header and not part of the actual metadata, are they?

Reply:
When reading from a USB frame they are in the header, when saved in a software_device part of the blob.
Anyway I don't think we should separate them to two messages, especially if it is only because of some LibRS implementation detail.

Reply:
I didn't say two messages. I meant:
`{{ ... header ... }, { ... metadata ... }}`
I.e. ,the same message will differentiate between actual frame metadata and what we added to it.

Reply:
Implemented as one message separated to header and metadata map fields